### ISSUE TITLE
AI improvements concerning player bases

### ISSUE THREAD

**Issue by Nitromachine** [beyond-all-reason/Beyond-All-Reason#2620]
### Description

I would like an option that forces friendly AI units and the player's own units to avoid moving through defenses that have been built by human players, such as dragon teeth or walls, instead of constantly destroying them. In addition, a second option that forces allied AI to avoid placing structures inside the general area of a human player's base(s).

### Why?

My biggest frustration with this game is having to deconstruct structures built by allied Barbarian AI or having to rebuild critical defenses because a Titan decided to take a shortcut through a wall and then crushing the rest of that same wall for no reason. I've had an AI ally build a single dragon tooth in the middle of an area that I left purposely empty for the output of my vehicle factories: 15 minutes after removing the tooth, the AI built it again in the same spot. I often have to deconstruct AI built T1 laser turrets because they are in the way of wind turbines or fusion reactors that I need to build. If an AI spawns next to me, I have to rush building metal extractors in certain spots or else the AI will take them. I don't want the AI building T1 radars in the middle of my base while an entire side of theirs is defenseless and lined with a chain of very explosive structures. I've had many more problems but you get the idea.

### What area of the game does this feature request pertain?

Gameplay, Settings, Controls or Hotkeys

### Other (optional)

This could perhaps be implemented as a circular zone of avoidance around structures built by human players, with bigger structures having a bigger radius. AI players can continue to ignore each others' boundaries to prevent the units of one AI from being boxed in by another friendly AI. Another possibility would be to enable players to draw zones where the AI shouldn't move or shouldn't build.

**Comment by PtaQQ** [comment]
Approved by the GDT, pending the concept of implementation.

**Comment by pandaro** [comment]
why me? im not able to work with barb :( i can try to do something on STAI, but it is not my prio at moment

**Comment by pandaro** [comment]
@WatchTheFort thanks for your trust anyway

**Comment by rlcevg** [comment]
An idea: when dragon teeth or walls finished mark it with UnitRulesParam "builderTeamId" or "builderAllyTeamId", because currently it's a neutral unit, so anyone can crash it.
Another option is to avoid any neutral unit while making path, but that's for prepayed AI version.

**Comment by rlcevg** [comment]
BARb, avoidance and scriptable ally zone size: https://github.com/beyond-all-reason/spring/commit/f3ba23635e1462ae2084f10bf9ba777467d16090
Requires testing.
Doubt someone will implement walls / neutral units avoidance. Won't expect it in a free version.

----

> Another possibility would be to enable players to draw zones where the AI shouldn't move or shouldn't build.

Is an interesting idea. Just like cooperation with AI through map marks. Waiting for UI implementation: presumably AI can receive drawn on a map polygons as a list of vertices encoded in a string (unfortunately) and process it in any way it wants internally: parse and map to a grid with specific cell size, or raw in-polygon test, whatever AI implements.

**Comment by eode** [comment]
I know this is old, but AI still plop buildings down right in the middle of my base.  Like -- abandon their spot, come over, upgrade my mex (which is ok I guess, but kinda sucks, because often enough I'd like to switch to T-1.5 mexes, but it's currently being upgraded by a friendly AI who isn't building his own base), and then he decides "Oh look, now that I'm here, this is a nice spot for a vehicle factory." 💢

**Comment by eode** [comment]
At least on glitters, I guess I can build a wall out of wind turbines to block 'em..

**Comment by saurtron** [comment]
for future reference, the commit where ally base avoidance was added https://github.com/rlcevg/CircuitAI/commit/90981974b64c9dc34190e0e1276cb5f2a91f1211

**Comment by Juksefantomet** [comment]
I can add to this.

I play a lot vs AI and in my 16vs16 games or even 32vs32 games.

The AI unintentionally block every path on smaller maps for allied and their own units.

Larger maps have enough room for this to not be as noticeable. 

But any map with a smaller size preferably narrow halls e.g `Asteroid Mines V3`

On this map they will build so many T1 windmills that they block their own and allied players ground units from being able to path to any location. I spent the better half of the playthrough on that map blowing up windmill farms blocking paths to remove them effectively enough 😂 

-

In short if they had a more sober approach to energy buildings e.g replacing some windmills with solar or avoid windmills and prioritize fusion++ they would most likely naturally avoid this behavior.
